### PR TITLE
KAFKA-13976: Improvements for OpenAPI specs

### PR DESCRIPTION
**Pull request by vamossagar12**
1) Added relevant APIResponses for all API endpoints.
2) Documented response codes/query-parameters/path-parameters/request body for all Connect related API endpoints.

### CODE REVIEW

**Comment by vamossagar12** [comment]
@mimaison Plz review. For some of the APIs returning Response, I have tried to use the actual class as response types. Also, there are a few endpoints, which return Maps. I have tried to use MapSchema as I couldn't find much documentation/examples around this. Do you think MapSchema would work? Thanks!

**Comment by vamossagar12** [comment]
@tombentley i saw the original PR for adding OpenAPI spec was reviewed by you as well. Plz review this one as well whenever you get the chance. Thanks!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java [hunk 1]
```diff
@@ -105,7 +115,28 @@ private <T> void addConnectorPlugins(Collection<PluginDesc<T>> plugins, Collecti
 
     @PUT
     @Path("/{pluginName}/config/validate")
-    @Operation(summary = "Validate the provided configuration against the configuration definition for the specified pluginName")
+    @Operation(
+            summary = "Validate the provided configuration against the configuration definition for the specified pluginName",
+            parameters = {
```
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java]: It looks like declaring parameters like this results in duplicates in the output, for example:

```yaml
  /connector-plugins/{pluginName}/config/validate:
    put:
      operationId: validateConfigs
      parameters:
      - in: path
        name: pluginName
        required: true
        schema:
          $ref: '#/components/schemas/StringSchema'
      - in: path
        name: pluginName
        required: true
        schema:
          type: string
```

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java [hunk 2]
```diff
@@ -105,7 +115,28 @@ private <T> void addConnectorPlugins(Collection<PluginDesc<T>> plugins, Collecti
...
+                    @Parameter(in = ParameterIn.PATH, schema = @Schema(implementation = StringSchema.class), name = "pluginName"),
```
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java]: `StringSchema` doesn't get rendered as a `type: string` as I'd expect, but rather as a `StringSchema` schema object with `type: object` and a whole bunch of schema properties. I think `@Schema(type = "string")` will give you what you need.

**Review by tombentley — COMMENTED** [review]
Thanks @vamossagar12, it would be nice to improve the generated schema. I left a couple of comments.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java (hunk 1 above)
Comment (vamossagar12) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java]: oh that's a good catch. I missed that.. Thanks for pointing it out. I removed the parameters from the operation annotation wherever the query param isn't hidden.

**Comment by vamossagar12** [comment]
Thanks for the review @tombentley . I made the changes as suggested

**Comment by vamossagar12** [comment]
@tombentley could you plz take a look whenever you get the chance. Thanks!

**Comment by vamossagar12** [comment]
@tombentley / @mimaison Plz review this PR whenever you get the chance..

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java [hunk 3]
```diff
@@ -112,7 +119,25 @@ public void requestTimeout(long requestTimeoutMs) {
...
+                            mediaType = "application/json",
+                            schema = @Schema(implementation = MapSchema.class),
+                            additionalPropertiesSchema = @Schema(type = "string")
+                    )
+            )
+    )
+    @ApiResponses(
+            value = {
+                    @ApiResponse(
+                            responseCode = "201",
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java:136]: Isn't it returning 200 instead?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java [hunk 4]
```diff
@@ -112,7 +119,25 @@ public void requestTimeout(long requestTimeoutMs) {
...
     @PUT
     @Path("/{pluginName}/config/validate")
-    @Operation(summary = "Validate the provided configuration against the configuration definition for the specified pluginName")
+    @Operation(
+            summary = "Validate the provided configuration against the configuration definition for the specified pluginName",
+            requestBody = @RequestBody(
+                    required = true,
+                    content = @Content(
+                            mediaType = "application/json",
+                            schema = @Schema(implementation = MapSchema.class),
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java:128]: This generates a ton of lines in the spec file which I'm not sure totally make sense. Have you tried using the spec file to generate code?

Same for all the other places where you put `MapSchema`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/LoggingResource.java [hunk 5]
```diff
@@ -128,7 +146,17 @@ public Response getLogger(final @PathParam("logger") String namedLogger) {
...
     @PUT
     @Path("/{logger}")
-    @Operation(summary = "Set the level for the specified logger")
+    @Operation(
+            summary = "Set the level for the specified logger",
+            requestBody = @RequestBody(
+                    required = true,
+                    content = @Content(
+                            mediaType = "application/json",
+                            schema = @Schema(implementation = MapSchema.class),
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/LoggingResource.java:155]: Should we cover the return type too? Currently it generates
```
responses:
        default:
          content:
            application/json: {}
          description: default response
```
which is not great

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 6]
```diff
@@ -107,7 +117,23 @@ public void requestTimeout(long requestTimeoutMs) {
 
     @GET
     @Path("/")
-    @Operation(summary = "List all active connectors")
+    @Operation(
+            summary = "List all active connectors",
+            parameters = @Parameter(
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:122]: Could we document this query parameter like we've done for `connectorsOnly`? See https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java#L148

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 7]
```diff
@@ -145,7 +171,26 @@ public Response listConnectors(
...
     @Path("/")
-    @Operation(summary = "Create a new connector")
+    @Operation(
+            summary = "Create a new connector",
+            requestBody = @RequestBody(
+                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = CreateConnectorRequest.class)),
+                    required = true,
+                    description = "Name and configuration parameters of the new connector. All values in config should be strings."
+            ),
+            parameters = @Parameter(in = ParameterIn.QUERY, hidden = true, schema = @Schema(implementation = BooleanSchema.class))
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:181]: Do we need this? I'm assuming it's about the `forward` query parameter but this does not even include the name! Also `forward` is already documented below.

Comment (mimaison) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:181]: Same a few more times below

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 8]
```diff
@@ -145,7 +171,26 @@ public Response listConnectors(
...
+    )
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "201",
+                    description = "New connector created successfully",
+                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = ConnectorInfo.class))
+            ),
+            @ApiResponse(
+                    responseCode = "409",
+                    description = "Error while creating new connector. Either rabalance is in progress or the connector already exists"
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:191]: Typo in `rebalance`. Same a few more times below.
Also can we rephrase it to: `Either a rebalance is in progress or the connector already exists`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 9]
```diff
@@ -179,7 +237,27 @@ public ConnectorInfo getConnector(final @PathParam("connector") String connector
...
+    @Operation(
+            summary = "Get the configuration for the specified connector",
+            parameters = {
+                    @Parameter(in = ParameterIn.QUERY, hidden = true, schema = @Schema(implementation = BooleanSchema.class), name = "forward")
+            }
+    )
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "200",
+                    description = "Connector config details retrieved successfully",
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:249]: Can we use `Connector configuration retrieved successfully`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 10]
```diff
@@ -188,9 +266,30 @@ public Map<String, String> getConnectorConfig(final @PathParam("connector") Stri
         return completeOrForwardRequest(cb, "/connectors/" + connector + "/config", "GET", headers, null, forward);
     }
 
+
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:269]: We don't need this extra line

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 11]
```diff
@@ -188,9 +266,30 @@ public Map<String, String> getConnectorConfig(final @PathParam("connector") Stri
...
+    @Operation(
+            summary = "Get the configuration of all tasks for the specified connector",
+            parameters = {
+                    @Parameter(in = ParameterIn.QUERY, hidden = true, schema = @Schema(implementation = BooleanSchema.class), name = "forward")
+            }
+    )
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "200",
+                    description = "Connector tasks config details retrieved successfully",
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:281]: What about `Connector tasks configuration retrieved successfully`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 12]
```diff
@@ -202,14 +301,39 @@ public Map<ConnectorTaskId, Map<String, String>> getTasksConfig(
...
         return herder.connectorStatus(connector);
     }
 
     @GET
     @Path("/{connector}/topics")
     @Operation(summary = "Get the list of topics actively used by the specified connector")
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "200",
+                    description = "Connector status retrieved successfully",
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:325]: It's not the right description

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 13]
```diff
@@ -222,6 +346,16 @@ public Response getConnectorActiveTopics(final @PathParam("connector") String co
     @PUT
     @Path("/{connector}/topics/reset")
     @Operation(summary = "Reset the list of topics actively used by the specified connector")
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "200",
+                    description = "Connector status retrieved successfully",
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:352]: Again not the right description

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 14]
```diff
@@ -237,7 +371,30 @@ public Response resetConnectorActiveTopics(final @PathParam("connector") String
...
+                            mediaType = "application/json",
+                            schema = @Schema(implementation = MapSchema.class),
+                            additionalPropertiesSchema = @Schema(type = "string")
+                    )
+            )
+    )
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "201",
+                    description = "New connector created successfully",
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:390]: What about `Connector created or updated successfully`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 15]
```diff
@@ -237,7 +371,30 @@ public Response resetConnectorActiveTopics(final @PathParam("connector") String
...
+                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = ConnectorInfo.class))
+            ),
+            @ApiResponse(
+                    responseCode = "409",
+                    description = "Error while creating new connector. Either rabalance is in progress or the connector already exists"
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:395]: This message is not correct. Let's use `Error while creating or updating connector. A rebalance is in progress`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 16]
```diff
@@ -237,7 +371,30 @@ public Response resetConnectorActiveTopics(final @PathParam("connector") String
...
+                    content = @Content(
+                            mediaType = "application/json",
+                            schema = @Schema(implementation = MapSchema.class),
+                            additionalPropertiesSchema = @Schema(type = "string")
+                    )
+            )
+    )
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "201",
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:389]: Is it 201 or 200?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 17]
```diff
@@ -260,7 +417,42 @@ public Response putConnectorConfig(final @PathParam("connector") String connecto
...
+                    responseCode = "204",
+                    description = "Operation succeeded, but there is no content in the response"
+            ),
+            @ApiResponse(
+                    responseCode = "404",
+                    description = "Named connector does not exist"
+            ),
+            @ApiResponse(
+                    responseCode = "409",
+                    description = "A rebalance is needed, forthcoming, or underway while restarting any of the Connector and/or Task objects"
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:447]: We should use original descriptions rather than copying from the Confluent docs!

Comment (mimaison) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:447]: If other descriptions have been copied, let's update them too

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 18]
```diff
@@ -299,15 +494,32 @@ public Response pauseConnector(@PathParam("connector") String connector, final @
...
+            summary = "List all tasks for the specified connector",
+            parameters = {
+                    @Parameter(in = ParameterIn.QUERY, hidden = true, schema = @Schema(implementation = BooleanSchema.class), name = "forward")
+            }
+    )
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "200",
+                    content = @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = TaskInfo.class))),
+                    description = "Connector resume request accepted"
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:519]: That's not the right description

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 19]
```diff
@@ -319,7 +531,11 @@ public List<TaskInfo> getTaskConfigs(final @PathParam("connector") String connec
 
     @POST
     @Path("/{connector}/tasks")
-    @Operation(hidden = true, summary = "This operation is only for inter-worker communications")
+    @Operation(
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:534]: We don't need to document the request body since it's not a public API. Let's undo this change.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 20]
```diff
@@ -332,7 +548,11 @@ public void putTaskConfigs(final @PathParam("connector") String connector,
 
     @PUT
     @Path("/{connector}/fence")
-    @Operation(hidden = true, summary = "This operation is only for inter-worker communications")
+    @Operation(
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:551]: Not a public API, let's undo too

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 21]
```diff
@@ -345,6 +565,14 @@ public void fenceZombies(final @PathParam("connector") String connector,
     @GET
     @Path("/{connector}/tasks/{task}/status")
     @Operation(summary = "Get the state of the specified task for the specified connector")
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "200",
+                    content = @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = ConnectorStateInfo.TaskState.class))),
+                    description = "Task status fetched successfully for specified connector"
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:572]: We've used `retrieved` everywhere so let's stick to that instead of introducing `fetched` here. We can also drop the `for specified connector` ending.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 22]
```diff
@@ -366,7 +600,19 @@ public void restartTask(final @PathParam("connector") String connector,
...
+            summary = "Delete the specified connector",
+            parameters = {
+                    @Parameter(in = ParameterIn.QUERY, schema = @Schema(implementation = BooleanSchema.class), name = "forward", hidden = true)
+            }
+    )
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "204", description = "Connector deleted successfully"),
+            @ApiResponse(
+                    responseCode = "409",
+                    description = "Error while deleting new connector. Either rabalance is in progress or the connector already exists"
```
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:613]: `or the connector already exists` does not seem quite right!

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. Quite a few things are still not quite right, see my comments.

**Comment by vamossagar12** [comment]
> Thanks for the PR. Quite a few things are still not quite right, see my comments.

Thanks @mimaison . Sorry about that. I see there are quite a few misses in the PR. Would correct them as you pointed.

**Comment by vamossagar12** [comment]
Converting to draft. Need some time before I get to this.

**Comment by mimaison** [comment]
@vamossagar12 Are you still planning to complete this item?

**Comment by vamossagar12** [comment]
@mimaison , yes sorry this completely missed my radar. Let me get to it some time next week.

**Comment by vamossagar12** [comment]
@mimaison , turns out I mayn't be able to spend time on this one. I will close this PR and unassigned the JIRA.
